airbnb/autoscaler
# cache ASG InstanceTypes

## Summary

Previously, Cluster Autoscaler cached Launch Configurations for ASGs but not Launch Templates; the impacts of this are that any ASG that uses a Launch Template and has 0 size would repeatedly hit the AWS API to get the details of the LT.  If you have enough empty ASGs this can cause API limits to hit.

I refactor the `launchConfigurationInstanceTypeCache` into a generic `instanceTypeExpirationStore`, which can hold data for either Launch Configurations or Launch Templates.  At the end of the day, this just maps the ASG id to the instance type associated with that ASG.

## Testing done
### Unit testing
- existing unit tests updated
- new unit tests added for the instance type cache
### Manual testing/evaluation in our clusters
First we tested behaviour on the `test-mc-c` cluster; no changes to scaling behaviour were observed.

Next we tested behaviour on `dev-a` so as to reduce the impact of making many AWS API calls.  We see that the number of API calls does go down, as desired (this graphs the number of `DescribeLaunchTemplateVersion` calls CA is making):

![Screen Shot 2021-04-28 at 4 39 06 PM](https://user-images.githubusercontent.com/3268445/116485657-9bc9d080-a840-11eb-8c3c-9725fd01b072.png)

At around 13:20, 300 ASGs were added to the cluster.  Using the unpatched version of CA, we make about 3500 `DescribeLaunchTemplateVersion` calls; at about 14:45, the patched version of CA was inserted into the cluster, which reduced the number of API calls to between 50-100/minute, which is well below our limits.

Other API calls made by CA increase slightly during this time period, but not a concerning amount.  This first graph is `DescribeAutoScalingGroups`, which are around 25/minute:

![Screen Shot 2021-04-28 at 4 45 29 PM](https://user-images.githubusercontent.com/3268445/116485873-2b6f7f00-a841-11eb-8b80-19070a7c8298.png)

The second graph is `DescribeTags`, which are around 1-300/minute, which is still well below our thresholds for these types of calls:

![Screen Shot 2021-04-28 at 4 46 31 PM](https://user-images.githubusercontent.com/3268445/116485940-58bc2d00-a841-11eb-8028-b014518d0145.png)

From a performance standpoint, this change also significantly improves CA performance (by about an order of magnitude):

![Screen Shot 2021-04-28 at 4 47 30 PM](https://user-images.githubusercontent.com/3268445/116486007-81dcbd80-a841-11eb-9087-2baf7072b113.png)

During the time the new version of CA was running, I observed the cluster scale up and down, so scaling behaviour appears to still be working correctly:

![Screen Shot 2021-04-28 at 4 50 54 PM](https://user-images.githubusercontent.com/3268445/116486211-ebf56280-a841-11eb-9a12-96ffd1b17ffc.png)

## Reviewers
@Ramyak @grampelberg @jiancheung
